R/mod_common_forms.R

```diff
@@ -93,60 +93,236 @@ mod_common_forms_server <- function(
+      reload_data(reload_data() + 1)
+      checked <- session$userData$update_checkboxes[[form]]
+      
+      df <- common_form_data() |> 
+        dplyr::mutate(o_reviewed = dplyr::if_else(subject_id == r$subject_id, 
+                                                  lapply(o_reviewed, modifyList, list(updated = checked)),
+                                                  o_reviewed))
+      common_form_data(df)
+      
       req(form == "Adverse events")
-      SAE_data <- data_active() |> 
-        dplyr::filter(grepl("Yes", `Serious Adverse Event`)) |> 
-        dplyr::select(dplyr::any_of(
-          c("subject_id","form_repeat", "Name", "AESI",  "SAE Start date", 
-            "SAE End date", "CTCAE severity", "Treatment related", 
-            "Treatment action", "Other action", "SAE Category", 
-            "SAE Awareness date", "SAE Date of death", "SAE Death reason")
-        )) |> 
-        adjust_colnames("^SAE ")
-      if(!input$show_all_data) SAE_data$subject_id <- NULL
-      datatable_custom(SAE_data, rename_vars = table_names, rownames= FALSE,
-                       title = "Serious Adverse Events", escape = FALSE)
+      df <- SAE_data() |> 
+        dplyr::mutate(o_reviewed = dplyr::if_else(subject_id == r$subject_id, 
+                                                  lapply(o_reviewed, modifyList, list(updated = checked)),
+                                                  o_reviewed))
+      SAE_data(df)
     })
     
-    output[["common_form_table"]] <- DT::renderDT({
-      df <- data_active()
-      if(form == "Adverse events") {
-        df <- df |> 
-          dplyr::filter(!grepl("Yes", `Serious Adverse Event`)
+    lapply(c("common_form", if (form == "Adverse events") "SAE"), \(tbl) {
+      # Reactive value mimics table name
+      tbl_data <- sprintf("%s_data", tbl)
+      # Get review selection input to observe
+      review_selection <- sprintf("%s_table_review_selection", tbl)
+      observeEvent(input[[review_selection]], {
+        # Update review values for session's user data
+        session$userData$update_checkboxes[[form]] <- NULL
+        session$userData$review_records[[form]] <-
+          dplyr::rows_upsert(
+            session$userData$review_records[[form]],
+            input[[review_selection]][c("id", "reviewed")],
+            by = "id"
+          ) |> 
+          dplyr::filter(!is.na(reviewed)) |> 
+          # Ensure that only the current subject is being reviewed
+          dplyr::semi_join(
+            subset(r$review_data, subject_id == r$subject_id & item_group == form),
+            by = "id"
           ) |> 
-          dplyr::select(-dplyr::starts_with("SAE"))
+          # Only update records where the review status is being changed
+          dplyr::anti_join(
+            subset(r$review_data, subject_id == r$subject_id & item_group == form),
+            by = c("id", "reviewed")
+          ) |> 
+          dplyr::arrange(id)
```

Comment:
the upsert/filter/semi_join and anti_join are a bit hard to follow for me. Is the semi_join really needed for example? Can't you just add subject_id and form to the filter directly?

Also, I think this part could be a candidate to extract in a helper function since it is duplicated in `mod_review_forms` and is more or less 'business logic'

Reply:
I will take a look.

Reply:
I cannot add `subject_id` and `item_group` to the filter directly because `input[[review_selection]]` does not contain those values. It only contains the `id`'s. This was intentional redundancy. Even though `session$userData$review_records[[form]]` gets reset whenever `r$subject_id` is changed, this ensures that no unintended records are being assessed. The overhead is very little because this is only triggered when a checkbox in a datatable is clicked.

Reply:
https://github.com/openpharma/clinsight/blob/f5fa833a2b5e9cd6caed609b27ede955d87b34b2/R/fct_form_helpers.R#L1-L39